### PR TITLE
[LLDB] Reapply SBSaveCore Add Memory List

### PR DESCRIPTION
**Pull request by Jlalond**
Recently in #107731 this change was revereted due to excess memory size in `TestSkinnyCore`. This was due to a bug where a range's end was being passed as size. Creating massive memory ranges.

Additionally, and requiring additional review, I added more unit tests and more verbose logic to the merging of save core memory regions.

@jasonmolenda as an FYI.

### CODE REVIEW

File: lldb/source/Target/CoreFileMemoryRanges.cpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "lldb/Target/CoreFileMemoryRanges.h"
+
+using namespace lldb;
+using namespace lldb_private;
+
+using Entry = CoreFileMemoryRanges::Entry;
```
Comment (Jlalond) [inline lldb/source/Target/CoreFileMemoryRanges.cpp:14]: @mbucko @clayborg can I get a rereview on this file. I added more logic.

File: lldb/include/lldb/Utility/RangeMap.h [hunk 2]
```diff
@@ -450,6 +450,8 @@ class RangeDataVector {
 
   void Append(const Entry &entry) { m_entries.emplace_back(entry); }
 
+  void Append(B &&b, S &&s, T &&t) { m_entries.emplace_back(Entry(b, s, t)); }
```
Comment (Jlalond) [inline lldb/include/lldb/Utility/RangeMap.h:453]: Nit to self, add code comments about Base Size And Data

**Review by clayborg — APPROVED** [review]

File: lldb/source/Target/CoreFileMemoryRanges.cpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+static bool IntersectHelper(const Entry *region_one, const Entry *region_two) {
+  return region_one->GetRangeBase() == region_two->GetRangeEnd() ||
+         region_one->GetRangeEnd() == region_two->GetRangeBase();
+}
+
+static bool OnlyIntersects(const Entry *region_one, const Entry *region_two) {
+  return IntersectHelper(region_one, region_two) ||
+         IntersectHelper(region_two, region_one);
```
Comment (mbucko) [inline lldb/source/Target/CoreFileMemoryRanges.cpp:28]: I believe `IntersectHelper(region_one, region_two)` is the same as `IntersectHelper(region_two, region_one)` no?

File: lldb/source/Target/CoreFileMemoryRanges.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+using namespace lldb_private;
+
+using Entry = CoreFileMemoryRanges::Entry;
+
+static bool Overlaps(const Entry *region_one, const Entry *region_two) {
+  return !(region_one->GetRangeEnd() < region_two->GetRangeBase() ||
+           region_two->GetRangeEnd() < region_one->GetRangeBase());
+}
+
+static bool IntersectHelper(const Entry *region_one, const Entry *region_two) {
```
Comment (mbucko) [inline lldb/source/Target/CoreFileMemoryRanges.cpp:21]: maybe a better name? for example `Touching()`

**Pull request merged**